### PR TITLE
Katerina_Podgornaya.11 node frameworks

### PR DESCRIPTION
**Pull request by ZZireal**

### CODE REVIEW

File: 11-node-frameworks/classworks/Podgornaya_Katerina/index.js [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+const router = new Router();
+app.keys = ['some-secret-key-one', 'some-secret-key-two'];
+const data = [];
+
+const schema = joi.object({
+    lastname: joi.string().required(),
+    description: joi.string().min(3).required(),
+});
+
+const get = (context) => {
```
Comment (anstill) [inline 11-node-frameworks/classworks/Podgornaya_Katerina/index.js]: keep in mind that counter should work for any request to the server, even posting form. Split counting and getting the "session counter"

Comment (anstill) [reply 11-node-frameworks/classworks/Podgornaya_Katerina/index.js]: basically it should be middleware which works for each request to the server, without any response

File: 11-node-frameworks/classworks/Podgornaya_Katerina/index.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+            context.body = `Some error happend!\n${error}`;
+        } else context.body = `Your data were got!`;
+    });
+
+app
+    .use(session())
+    .use(bodyParser())
+    .use(serve((`${__dirname}/static`)))
+    .use(router.routes())
+    .use((ctx) => ctx.body = '1')
```
Comment (anstill) [inline 11-node-frameworks/classworks/Podgornaya_Katerina/index.js]: Looks like a hack, and I'm not sure what the purpose for this

File: 11-node-frameworks/classworks/Podgornaya_Katerina/index.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+const Router = require('koa-router');
+const serve = require('koa-static');
+const bodyParser = require('koa-bodyparser');
+const joi = require('joi');
+let session = require('koa-generic-session');
+
+const app = new Koa();
+const router = new Router();
+app.keys = ['some-secret-key-one', 'some-secret-key-two'];
+const data = [];
```
Comment (anstill) [inline 11-node-frameworks/classworks/Podgornaya_Katerina/index.js]: use session as a storage

File: 11-node-frameworks/classworks/Podgornaya_Katerina/index.js [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        remove(context);
+    })
+    .get('/regenerate', async (context, next) => {
+        await regenerate(context);
+    })
+    .post('/data', async (context, next) => {
+        data.push(context.request.body);
+        console.log(data);
+        const { value, error } = schema.validate(context.request.body, { allowUnknown: true, abortEarly: false });
+        if (error) {
```
Comment (anstill) [inline 11-node-frameworks/classworks/Podgornaya_Katerina/index.js]: I like error handling 👍

Could you send them to client as well?

File: 11-node-frameworks/classworks/Podgornaya_Katerina/index.js [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+            context.body = `Some error happend!\n${error}`;
```
Comment (anstill) [inline 11-node-frameworks/classworks/Podgornaya_Katerina/index.js]: Please,  use JSON everywhere

File: 11-node-frameworks/classworks/Podgornaya_Katerina/index.js [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+const regenerate = async (context) => {
+    get(context);
+    await context.regenerateSession();
+    get(context);
+};
+
+router
+    .get('/get', (context, next) => {
+        get(context);
```
Comment (anstill) [inline 11-node-frameworks/classworks/Podgornaya_Katerina/index.js]: This endpoint should return data from a form as well

File: 11-node-frameworks/classworks/Podgornaya_Katerina/index.js [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+};
+
+const remove = (context) => {
+    context.session = null;
+    context.body = 0;
+};
+
+const regenerate = async (context) => {
+    get(context);
+    await context.regenerateSession();
```
Comment (anstill) [inline 11-node-frameworks/classworks/Podgornaya_Katerina/index.js]: 👍

**Review by anstill — CHANGES_REQUESTED** [review]
Hey, @ZZireal great job here! Please check my comments and feel free to ask any questions.

File: 11-node-frameworks/classworks/Podgornaya_Katerina/index.js [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+    const { value, error } = schema.validate(context.request.body, {
+      allowUnknown: true, abortEarly: false,
+    });
+    if (error) {
+      context.body += `\nError happend: ${JSON.stringify(error.message)}.`;
+    } else {
+      if (!session.state) session.state = [];
+      session.state.push(context.request.body);
+      context.body = JSON.stringify(session.state);
+      context.body += JSON.stringify('Your data were got!');
```
Comment (anstill) [inline 11-node-frameworks/classworks/Podgornaya_Katerina/index.js]: if you add JSON + JSON it isn't a valid JSON as a result. 

Please, don't use `+=` construction for setting response

File: 11-node-frameworks/classworks/Podgornaya_Katerina/index.js (hunk 1 above)
Comment (anstill) [reply 11-node-frameworks/classworks/Podgornaya_Katerina/index.js]: Still valid point

**Review by anstill — CHANGES_REQUESTED** [review]

File: 11-node-frameworks/classworks/Podgornaya_Katerina/index.js (hunk 1 above)
Comment (ZZireal) [reply 11-node-frameworks/classworks/Podgornaya_Katerina/index.js]: i added middleware before routing, so for every request to server counter works and counter number you can see at '/get' request (now you can see counter number at every request, i add it to every ctx.body):

my middleware:
.use(async (context, next) => {
    setCounter(context);
    await next();
  })

is it wrong?

image

**Review by anstill — APPROVED** [review]

**Pull request merged**